File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/node/CrafterNetworkNode.java
package com.refinedmods.refinedstorage.apiimpl.network.node;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPattern;
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPatternContainer;
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPatternProvider;
import com.refinedmods.refinedstorage.api.network.INetwork;
import com.refinedmods.refinedstorage.api.network.node.INetworkNode;
import com.refinedmods.refinedstorage.apiimpl.API;
import com.refinedmods.refinedstorage.inventory.item.BaseItemHandler;
import com.refinedmods.refinedstorage.inventory.item.UpgradeItemHandler;
import com.refinedmods.refinedstorage.inventory.item.validator.PatternItemValidator;
import com.refinedmods.refinedstorage.inventory.listener.NetworkNodeInventoryListener;
import com.refinedmods.refinedstorage.item.UpgradeItem;
import com.refinedmods.refinedstorage.util.LevelUtils;
import com.refinedmods.refinedstorage.util.StackUtils;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.Nameable;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraftforge.fluids.capability.IFluidHandler;
import net.minecraftforge.items.IItemHandler;
import net.minecraftforge.items.IItemHandlerModifiable;
import net.minecraftforge.items.wrapper.CombinedInvWrapper;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public class CrafterNetworkNode extends NetworkNode implements ICraftingPatternContainer {

    public static final ResourceLocation ID = new ResourceLocation(RS.ID, "crafter");
    private static final Component DEFAULT_NAME = Component.translatable("gui.refinedstorage.crafter");
    private static final String NBT_DISPLAY_NAME = "DisplayName";
    private static final String NBT_UUID = "CrafterUuid";
    private static final String NBT_MODE = "Mode";
    private static final String NBT_LOCKED = "Locked";
    private static final String NBT_WAS_POWERED = "WasPowered";
    private final List<ICraftingPattern> patterns = new ArrayList<>();
    private final UpgradeItemHandler upgrades = (UpgradeItemHandler) new UpgradeItemHandler(4, UpgradeItem.Type.SPEED)
        .addListener(new NetworkNodeInventoryListener(this));
    // Used to prevent infinite recursion on getRootContainer() when there's e.g. two crafters facing each other.
    private boolean visited = false;
    private final BaseItemHandler patternsInventory = new BaseItemHandler(9) {
        @Override
        public int getSlotLimit(int slot) {
            return 1;
        }

        @Nonnull
        @Override
        public ItemStack insertItem(int slot, @Nonnull ItemStack stack, boolean simulate) {
            if (!stacks.get(slot).isEmpty()) {
                return stack;
            }

            return super.insertItem(slot, stack, simulate);
        }
    }
        .addValidator(new PatternItemValidator(level))
        .addListener(new NetworkNodeInventoryListener(this))
        .addListener((handler, slot, reading) -> {
            if (!reading) {
                if (!level.isClientSide) {
                    invalidate();
                }

                if (network != null) {
                    network.getCraftingManager().invalidate();
                }
            }
        });
    private CrafterMode mode = CrafterMode.IGNORE;
    private boolean locked = false;
    private boolean wasPowered;
    @Nullable
    private Component displayName;
    @Nullable
    private UUID uuid = null;

    public CrafterNetworkNode(Level level, BlockPos pos) {
        super(level, pos);
    }

    private void invalidate() {
        patterns.clear();

        for (int i = 0; i < patternsInventory.getSlots(); ++i) {
            ItemStack patternStack = patternsInventory.getStackInSlot(i);

            if (!patternStack.isEmpty()) {
                ICraftingPattern pattern = ((ICraftingPatternProvider) patternStack.getItem()).create(level, patternStack, this);

                if (pattern.isValid()) {
                    patterns.add(pattern);
                }
            }
        }
    }

    @Override
    public int getEnergyUsage() {
        return RS.SERVER_CONFIG.getCrafter().getUsage() + upgrades.getEnergyUsage() + (RS.SERVER_CONFIG.getCrafter().getPatternUsage() * patterns.size());
    }

    @Override
    public void update() {
        super.update();

        if (ticks == 1) {
            invalidate();
        }

        if (mode == CrafterMode.PULSE_INSERTS_NEXT_SET && level.isLoaded(pos)) {
            if (level.hasNeighborSignal(pos)) {
                this.wasPowered = true;

                markDirty();
            } else if (wasPowered) {
                this.wasPowered = false;
                this.locked = false;

                markDirty();
            }
        }
    }

    @Override
    protected void onConnectedStateChange(INetwork network, boolean state, ConnectivityStateChangeCause cause) {
        super.onConnectedStateChange(network, state, cause);

        network.getCraftingManager().invalidate();
    }

    @Override
    public void onDisconnected(INetwork network) {
        super.onDisconnected(network);

        network.getCraftingManager().getTasks().stream()
            .filter(task -> task.getPattern().getContainer().getPosition().equals(pos))
            .forEach(task -> network.getCraftingManager().cancel(task.getId()));
    }

    @Override
    public void onDirectionChanged(Direction direction) {
        super.onDirectionChanged(direction);

        if (network != null) {
            network.getCraftingManager().invalidate();
        }
    }

    @Override
    public void read(CompoundTag tag) {
        super.read(tag);

        StackUtils.readItems(patternsInventory, 0, tag);

        invalidate();

        StackUtils.readItems(upgrades, 1, tag);

        if (tag.contains(NBT_DISPLAY_NAME)) {
            displayName = Component.Serializer.fromJson(tag.getString(NBT_DISPLAY_NAME));
        }

        if (tag.hasUUID(NBT_UUID)) {
            uuid = tag.getUUID(NBT_UUID);
        }

        if (tag.contains(NBT_MODE)) {
            mode = CrafterMode.getById(tag.getInt(NBT_MODE));
        }

        if (tag.contains(NBT_LOCKED)) {
            locked = tag.getBoolean(NBT_LOCKED);
        }

        if (tag.contains(NBT_WAS_POWERED)) {
            wasPowered = tag.getBoolean(NBT_WAS_POWERED);
        }
    }

    @Override
    public ResourceLocation getId() {
        return ID;
    }

    @Override
    public CompoundTag write(CompoundTag tag) {
        super.write(tag);

        StackUtils.writeItems(patternsInventory, 0, tag);
        StackUtils.writeItems(upgrades, 1, tag);

        if (displayName != null) {
            tag.putString(NBT_DISPLAY_NAME, Component.Serializer.toJson(displayName));
        }

        if (uuid != null) {
            tag.putUUID(NBT_UUID, uuid);
        }

        tag.putInt(NBT_MODE, mode.ordinal());
        tag.putBoolean(NBT_LOCKED, locked);
        tag.putBoolean(NBT_WAS_POWERED, wasPowered);

        return tag;
    }

    @Override
    public int getUpdateInterval() {
        switch (upgrades.getUpgradeCount(UpgradeItem.Type.SPEED)) {
            case 0:
                return 10;
            case 1:
                return 8;
            case 2:
                return 6;
            case 3:
                return 4;
            case 4:
                return 2;
            default:
                return 0;
        }
    }

    @Override
    public int getMaximumSuccessfulCraftingUpdates() {
        switch (upgrades.getUpgradeCount(UpgradeItem.Type.SPEED)) {
            case 0:
                return 1;
            case 1:
                return 2;
            case 2:
                return 3;
            case 3:
                return 4;
            case 4:
                return 5;
            default:
                return 1;
        }
    }

    @Override
    @Nullable
    public IItemHandler getConnectedInventory() {
        ICraftingPatternContainer proxy = getRootContainer();
        if (proxy == null) {
            return null;
        }

        return LevelUtils.getItemHandler(proxy.getFacingBlockEntity(), proxy.getDirection().getOpposite());
    }

    @Nullable
    @Override
    public IFluidHandler getConnectedFluidInventory() {
        ICraftingPatternContainer proxy = getRootContainer();
        if (proxy == null) {
            return null;
        }

        return LevelUtils.getFluidHandler(proxy.getFacingBlockEntity(), proxy.getDirection().getOpposite());
    }

    @Override
    @Nullable
    public BlockEntity getConnectedBlockEntity() {
        ICraftingPatternContainer proxy = getRootContainer();
        if (proxy == null) {
            return null;
        }

        return proxy.getFacingBlockEntity();
    }

    @Override
    public List<ICraftingPattern> getPatterns() {
        return patterns;
    }

    @Override
    @Nullable
    public IItemHandlerModifiable getPatternInventory() {
        return patternsInventory;
    }

    @Override
    public Component getName() {
        if (displayName != null) {
            return displayName;
        }

        BlockEntity facing = getConnectedBlockEntity();

        if (facing instanceof Nameable && ((Nameable) facing).getName() != null) {
            return ((Nameable) facing).getName();
        }

        if (facing != null) {
            return Component.translatable(level.getBlockState(facing.getBlockPos()).getBlock().getDescriptionId());
        }

        return DEFAULT_NAME;
    }

    @Nullable
    public Component getDisplayName() {
        return displayName;
    }

    public void setDisplayName(Component displayName) {
        this.displayName = displayName;
    }

    @Override
    public BlockPos getPosition() {
        return pos;
    }

    public CrafterMode getMode() {
        return mode;
    }

    public void setMode(CrafterMode mode) {
        this.mode = mode;
        this.wasPowered = false;
        this.locked = false;

        this.markDirty();
    }

    public IItemHandler getUpgrades() {
        return upgrades;
    }

    @Override
    public IItemHandler getDrops() {
        return new CombinedInvWrapper(patternsInventory, upgrades);
    }

    @Override
    @Nullable
    public ICraftingPatternContainer getRootContainer() {
        if (visited) {
            return null;
        }

        INetworkNode facing = API.instance().getNetworkNodeManager((ServerLevel) level).getNode(pos.relative(getDirection()));
        if (!(facing instanceof ICraftingPatternContainer) || facing.getNetwork() != network) {
            return this;
        }

        visited = true;
        ICraftingPatternContainer facingContainer = ((ICraftingPatternContainer) facing).getRootContainer();
        visited = false;

        return facingContainer;
    }

    public Optional<ICraftingPatternContainer> getRootContainerNotSelf() {
        ICraftingPatternContainer root = getRootContainer();

        if (root != null && root != this) {
            return Optional.of(root);
        }

        return Optional.empty();
    }

    @Override
    public UUID getUuid() {
        if (this.uuid == null) {
            this.uuid = UUID.randomUUID();

            markDirty();
        }

        return uuid;
    }

    @Override
    public boolean isLocked() {
        Optional<ICraftingPatternContainer> root = getRootContainerNotSelf();
        if (root.isPresent()) {
            return root.get().isLocked();
        }

        switch (mode) {
            case IGNORE:
                return false;
            case SIGNAL_LOCKS_AUTOCRAFTING:
                return level.hasNeighborSignal(pos);
            case SIGNAL_UNLOCKS_AUTOCRAFTING:
                return !level.hasNeighborSignal(pos);
            case PULSE_INSERTS_NEXT_SET:
                return locked;
            default:
                return false;
        }
    }

    @Override
    public void unlock() {
        locked = false;
    }

    @Override
    public void onUsedForProcessing() {
        Optional<ICraftingPatternContainer> root = getRootContainerNotSelf();
        if (root.isPresent()) {
            root.get().onUsedForProcessing();

            return;
        }

        if (mode == CrafterMode.PULSE_INSERTS_NEXT_SET) {
            this.locked = true;

            markDirty();
        }
    }

    public enum CrafterMode {
        IGNORE,
        SIGNAL_UNLOCKS_AUTOCRAFTING,
        SIGNAL_LOCKS_AUTOCRAFTING,
        PULSE_INSERTS_NEXT_SET;

        public static CrafterMode getById(int id) {
            if (id >= 0 && id < values().length) {
                return values()[id];
            }

            return IGNORE;
        }
    }


}

File: src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/node/CrafterNetworkNode.kt
package com.refinedmods.refinedstorage.apiimpl.network.node

import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPattern
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPatternContainer
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPatternProvider
import com.refinedmods.refinedstorage.api.autocrafting.task.ICraftingTask
import com.refinedmods.refinedstorage.api.network.INetwork
import com.refinedmods.refinedstorage.api.network.node.INetworkNode
import com.refinedmods.refinedstorage.apiimpl.API
import com.refinedmods.refinedstorage.apiimpl.network.nodeimport.ConnectivityStateChangeCause
import com.refinedmods.refinedstorage.inventory.item.BaseItemHandler
import com.refinedmods.refinedstorage.inventory.item.UpgradeItemHandler
import com.refinedmods.refinedstorage.inventory.item.validator.PatternItemValidator
import com.refinedmods.refinedstorage.inventory.listener.InventoryListener
import com.refinedmods.refinedstorage.inventory.listener.NetworkNodeInventoryListener
import com.refinedmods.refinedstorage.item.UpgradeItem
import com.refinedmods.refinedstorage.util.StackUtils
import com.refinedmods.refinedstorage.util.WorldUtils
import net.minecraft.block.entity.BlockEntity
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraft.server.world.ServerWorld
import net.minecraft.text.Text
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import java.util.*
import java.util.function.Consumer
import java.util.function.Predicate

class CrafterNetworkNode(world: World, pos: BlockPos?) : NetworkNode(world, pos), ICraftingPatternContainer {
    enum class CrafterMode {
        IGNORE, SIGNAL_UNLOCKS_AUTOCRAFTING, SIGNAL_LOCKS_AUTOCRAFTING, PULSE_INSERTS_NEXT_SET;

        companion object {
            @kotlin.jvm.JvmStatic
            fun getById(id: Int): CrafterMode {
                return if (id >= 0 && id < values().size) {
                    values()[id]
                } else IGNORE
            }
        }
    }

    private val patternsInventory: BaseItemHandler = object : BaseItemHandler(9) {
        fun getSlotLimit(slot: Int): Int {
            return 1
        }
    }
            .addValidator(PatternItemValidator(world))
            .addListener(NetworkNodeInventoryListener(this))
            .addListener(InventoryListener<BaseItemHandler> { handler: BaseItemHandler?, slot: Int, reading: Boolean ->
                if (!reading) {
                    if (!world.isClient) {
                        invalidate()
                    }
                    if (network != null) {
                        network!!.craftingManager.invalidate()
                    }
                }
            })
    private val patterns: MutableList<ICraftingPattern?> = ArrayList<ICraftingPattern?>()
    private val upgrades: UpgradeItemHandler = UpgradeItemHandler(4, UpgradeItem.Type.SPEED)
            .addListener(NetworkNodeInventoryListener(this)) as UpgradeItemHandler

    // Used to prevent infinite recursion on getRootContainer() when there's e.g. two crafters facing each other.
    private var visited = false
    private var mode = CrafterMode.IGNORE
    private var locked = false
    private var wasPowered = false

    var displayName: Text? = null

    @Nullable
    private var uuid: UUID? = null
    private fun invalidate() {
        patterns.clear()
        for (i in 0 until patternsInventory.getSlots()) {
            val patternStack: ItemStack = patternsInventory.getStackInSlot(i)
            if (!patternStack.isEmpty()) {
                val pattern: ICraftingPattern = (patternStack.getItem() as ICraftingPatternProvider).create(world, patternStack, this)
                if (pattern.isValid()) {
                    patterns.add(pattern)
                }
            }
        }
    }

    override val energyUsage: Int
        get() = RS.SERVER_CONFIG.getCrafter().getUsage() + upgrades.getEnergyUsage() + RS.SERVER_CONFIG.getCrafter().getPatternUsage() * patterns.size

    override fun update() {
        super.update()
        if (ticks == 1) {
            invalidate()
        }
        if (mode == CrafterMode.PULSE_INSERTS_NEXT_SET && world.isBlockPresent(pos)) {
            if (world.isReceivingRedstonePower(pos)) {
                wasPowered = true
                markDirty()
            } else if (wasPowered) {
                wasPowered = false
                locked = false
                markDirty()
            }
        }
    }

    protected override fun onConnectedStateChange(network: INetwork?, state: Boolean, cause: ConnectivityStateChangeCause?) {
        super.onConnectedStateChange(network, state, cause)
        network.craftingManager.invalidate()
    }

    override fun onDisconnected(network: INetwork?) {
        super.onDisconnected(network)
        network.craftingManager.getTasks().stream()
                .filter(Predicate<ICraftingTask> { task: ICraftingTask -> task.getPattern().getContainer().getPosition() == pos })
                .forEach(Consumer<ICraftingTask> { task: ICraftingTask -> network.craftingManager.cancel(task.getId()) })
    }

    fun onDirectionChanged(direction: Direction?) {
        super.onDirectionChanged(direction)
        if (network != null) {
            network!!.craftingManager.invalidate()
        }
    }

    override fun read(tag: CompoundTag) {
        super.read(tag)
        StackUtils.readItems(patternsInventory, 0, tag)
        invalidate()
        StackUtils.readItems(upgrades, 1, tag)
        if (tag.contains(NBT_DISPLAY_NAME)) {
            displayName = Text.Serializer.func_240643_a_(tag.getString(NBT_DISPLAY_NAME))
        }
        if (tag.hasUniqueId(NBT_UUID)) {
            uuid = tag.getUniqueId(NBT_UUID)
        }
        if (tag.contains(NBT_MODE)) {
            mode = CrafterMode.getById(tag.getInt(NBT_MODE))
        }
        if (tag.contains(NBT_LOCKED)) {
            locked = tag.getBoolean(NBT_LOCKED)
        }
        if (tag.contains(NBT_WAS_POWERED)) {
            wasPowered = tag.getBoolean(NBT_WAS_POWERED)
        }
    }

    override val id: Identifier
        get() = ID

    override fun write(tag: CompoundTag): CompoundTag {
        super.write(tag)
        StackUtils.writeItems(patternsInventory, 0, tag)
        StackUtils.writeItems(upgrades, 1, tag)
        if (displayName != null) {
            tag.putString(NBT_DISPLAY_NAME, Text.Serializer.toJson(displayName))
        }
        if (uuid != null) {
            tag.putUniqueId(NBT_UUID, uuid)
        }
        tag.putInt(NBT_MODE, mode.ordinal)
        tag.putBoolean(NBT_LOCKED, locked)
        tag.putBoolean(NBT_WAS_POWERED, wasPowered)
        return tag
    }

    val updateInterval: Int
        get() = when (upgrades.getUpgradeCount(UpgradeItem.Type.SPEED)) {
            0 -> 10
            1 -> 8
            2 -> 6
            3 -> 4
            4 -> 2
            else -> 0
        }
    val maximumSuccessfulCraftingUpdates: Int
        get() = when (upgrades.getUpgradeCount(UpgradeItem.Type.SPEED)) {
            0 -> 1
            1 -> 2
            2 -> 3
            3 -> 4
            4 -> 5
            else -> 1
        }

    @get:Nullable
    val connectedInventory: IItemHandler?
        get() {
            val proxy: ICraftingPatternContainer = rootContainer ?: return null
            return WorldUtils.getItemHandler(proxy.getFacingTile(), proxy.getDirection().getOpposite())
        }

    @get:Nullable
    val connectedFluidInventory: IFluidHandler?
        get() {
            val proxy: ICraftingPatternContainer = rootContainer ?: return null
            return WorldUtils.getFluidHandler(proxy.getFacingTile(), proxy.getDirection().getOpposite())
        }

    @get:Nullable
    val connectedTile: BlockEntity?
        get() {
            val proxy: ICraftingPatternContainer = rootContainer ?: return null
            return proxy.getFacingTile()
        }

    @get:Nullable
    override val facingTile: BlockEntity?
        get() {
            val facingPos: BlockPos = pos.offset(direction)
            return if (!world.isBlockPresent(facingPos)) {
                null
            } else world.getBlockEntity(facingPos)
        }

    override fun getPatterns(): List<ICraftingPattern?>? {
        return patterns
    }

    @get:Nullable
    val patternInventory: IItemHandlerModifiable?
        get() = patternsInventory
    val name: Text?
        get() {
            if (displayName != null) {
                return displayName
            }
            val facing: BlockEntity? = connectedTile
            if (facing is INameable && (facing as INameable?).getName() != null) {
                return (facing as INameable?).getName()
            }
            return if (facing != null) {
                TranslationTextComponent(world.getBlockState(facing.getPos()).block.translationKey)
            } else DEFAULT_NAME
        }

    fun setDisplayName(displayName: Text?) {
        this.displayName = displayName
    }

    val position: BlockPos?
        get() = pos

    fun getMode(): CrafterMode {
        return mode
    }

    fun setMode(mode: CrafterMode) {
        this.mode = mode
        wasPowered = false
        locked = false
        markDirty()
    }

    val patternItems: IItemHandler
        get() = patternsInventory

    fun getUpgrades(): IItemHandler {
        return upgrades
    }

    override val drops: IItemHandler
        get() = CombinedInvWrapper(patternsInventory, upgrades)

    @get:Nullable
    val rootContainer: ICraftingPatternContainer?
        get() {
            if (visited) {
                return null
            }
            val facing: INetworkNode = API.instance().getNetworkNodeManager(world as ServerWorld).getNode(pos.offset(direction))
            if (facing !is ICraftingPatternContainer || facing.network !== network) {
                return this
            }
            visited = true
            val facingContainer: ICraftingPatternContainer = (facing as ICraftingPatternContainer).getRootContainer()
            visited = false
            return facingContainer
        }
    val rootContainerNotSelf: Optional<Any>
        get() {
            val root: ICraftingPatternContainer? = rootContainer
            return if (root != null && root !== this) {
                Optional.of(root)
            } else Optional.empty()
        }

    override fun getUuid(): UUID? {
        if (uuid == null) {
            uuid = UUID.randomUUID()
            markDirty()
        }
        return uuid
    }

    override fun isLocked(): Boolean {
        val root: Optional<ICraftingPatternContainer> = rootContainerNotSelf
        return if (root.isPresent()) {
            root.get().isLocked()
        } else when (mode) {
            CrafterMode.IGNORE -> false
            CrafterMode.SIGNAL_LOCKS_AUTOCRAFTING -> world.isReceivingRedstonePower(pos)
            CrafterMode.SIGNAL_UNLOCKS_AUTOCRAFTING -> !world.isReceivingRedstonePower(pos)
            CrafterMode.PULSE_INSERTS_NEXT_SET -> locked
            else -> false
        }
    }

    override fun unlock() {
        locked = false
    }

    override fun onUsedForProcessing() {
        val root: Optional<ICraftingPatternContainer> = rootContainerNotSelf
        if (root.isPresent()) {
            root.get().onUsedForProcessing()
            return
        }
        if (mode == CrafterMode.PULSE_INSERTS_NEXT_SET) {
            locked = true
            markDirty()
        }
    }

    companion object {
        @kotlin.jvm.JvmField
        val ID: Identifier = Identifier(RS.ID, "crafter")
        private val DEFAULT_NAME: Text = TranslationTextComponent("gui.refinedstorage.crafter")
        private const val NBT_DISPLAY_NAME = "DisplayName"
        private const val NBT_UUID = "CrafterUuid"
        private const val NBT_MODE = "Mode"
        private const val NBT_LOCKED = "Locked"
        private const val NBT_WAS_POWERED = "WasPowered"
    }
}